spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -1062,6 +1085,84 @@ public Message<O> process(Message<I> msg) {
+As for the `POST_PROCESSING` positive acknowledgement strategy, a _nack_ is propagated from a message to its source if the messages are chained.
+When the _nack_ reaches the top of the chain, the source, generally a connector, can handle the failure and act accordingly.
+Typically, sending the invalid message to a dead-letter-queue or retrying the delivery are common strategies.
+
+Negative acknowledgement happens automatically if a method annotated with `@Incoming` receives a single payload and uses `POST_PROCESSING` acknowledgement and throws an exception.
+The incoming message is _nacked_ and the thrown exception is passed as reason.
+The following method would automatically trigger a `nack` if the incoming payload is `b`:
+
+[source, java]
+----
+@Incoming("data")
+@Outgoing("out")
+public String process(String s) {  // <1>
+    if (s.equalsIgnoreCase("b")) {
+        throw new IllegalArgumentException("b"); // <2>
+    }
+    return s.toUpperCase();
+}
+----
+1. `POST_PROCESSING` is the default strategy for this signature
+2. throwing an exception _nacks_ the incoming message
+
+If a method handles payload asynchronously (returns a `CompletionStage`), the incoming message is nacked if:
+
+* the method throws an exception
+* the produced `CompletionStage` is completed exceptionally
+
+[source, java]
+----
+@Incoming("data")
+@Outgoing("out")
+public CompletionStage<String> process(String s) {   // <1>
+    if (s.equalsIgnoreCase("b")) {
+        throw new IllegalArgumentException("b");     // <2>
+    }
+
+    if (s.equalsIgnoreCase("f")) {
+        return null;                                // <3>
+    }
+
+    if (s.equalsIgnoreCase("c")) {
+        CompletableFuture<String> cf = new CompletableFuture<>();
+        cf.completeExceptionally(new IllegalArgumentException("c"));  // <4>
+        return cf;
+    }
+    return CompletableFuture.completedFuture(s.toUpperCase());
+}
+----
+1. `POST_PROCESSING` is the default strategy for this signature
+2. throwing an exception _nacks_ the incoming message
+3. returning `null` nacks the incoming message, as it's illegal to return `null` as `CompletionStage`
+4. returning a `CompletionStage` failed exceptionally also nacks the incoming message
+
+If the method receives a single `Message`, the user is responsible for calling the `nack` function:
+
+[source, java]
+----
+@Incoming("data")
+@Outgoing("out")
+public Message<String> process(Message<String> m) {  // <1>
+    String s = m.getPayload();
+    if (s.equalsIgnoreCase("b")) {
+        // we cannot fail, we must nack explicitly.
+        m.nack(new IllegalArgumentException("b"));  // <2>
+        return null;
+    }
+    return m.withPayload(s.toUpperCase());
+}
+----
+1. `MANUAL` is the default strategy for this signature
+2. throwing an exception would not _nack_ the message, as it may have already been explicetely _acked_ or _nacked_, before the exception happens, so the user must call the `nack` method
```

Comment:
```suggestion
2. throwing an exception would not _nack_ the message, as it may have already been explicitly _acked_ or _nacked_. Before the exception happens, the user must call the `nack` method.
```